### ISSUE TITLE
Using edge compiler for middleware broke my middleware

### ISSUE THREAD

**Issue by TheThirdRace** [vercel/next.js#37823]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

[09:45] /d/_dev/thethirdrace.com (chore/update-dependencies)
$ yarn run next info
yarn run v1.22.17
$ D:\_dev\thethirdrace.com\node_modules\.bin\next info

    Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 16.15.1
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 12.1.6
      react: 18.2.0
      react-dom: 18.2.0

warn  - Latest canary version not detected, detected: "12.1.6", newest: "12.1.7-canary.41".
        Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
        Read more - https://nextjs.org/docs/messages/opening-an-issue
Done in 0.61s.

### What browser are you using? (if relevant)

Not relevant

### How are you deploying your application? (if relevant)

not relevant

### Describe the Bug

Upgrade NextJs from `12.1.5` to `12.1.6`.

Middleware use `@supabase/gotrue-js (1.22.16`) and `@supabase/postgrest-js (0.37.3)` to make some calls to the database.

Run `yarn build` after upgrade and get this error:

```shell
[09:49] /d/_dev/thethirdrace.com (chore/update-dependencies)
$ yarn build
yarn run v1.22.17
$ next build
info  - Loaded env from D:\_dev\thethirdrace.com\.env.local
info  - Skipping validation of types
warn  - using beta Middleware (not covered by semver) - https://nextjs.org/docs/messages/beta-middleware
info  - Creating an optimized production build
Failed to compile.

./node_modules/next/dist/compiled/native-url/index.js
Module not found: Can't resolve 'querystring' in 'D:\_dev\thethirdrace.com\node_modules\next\dist\compiled\native-url'

Import trace for requested module:
./node_modules/node-fetch/lib/index.mjs
./node_modules/cross-fetch/dist/node-ponyfill.js
./node_modules/@supabase/postgrest-js/dist/module/lib/types.js
./node_modules/@supabase/postgrest-js/dist/module/index.js
./src/helper/Database.ts
./src/pages/api/payment/_middleware.api.ts

./node_modules/whatwg-url/lib/url-state-machine.js
Module not found: Can't resolve 'buffer' in 'D:\_dev\thethirdrace.com\node_modules\whatwg-url\lib'

Import trace for requested module:
./node_modules/whatwg-url/lib/public-api.js
./node_modules/node-fetch/lib/index.mjs
./node_modules/cross-fetch/dist/node-ponyfill.js
./node_modules/@supabase/postgrest-js/dist/module/lib/types.js
./node_modules/@supabase/postgrest-js/dist/module/index.js
./src/helper/Database.ts
./src/pages/api/payment/_middleware.api.ts

./node_modules/node-fetch/lib/index.mjs
Module not found: Can't resolve 'stream' in 'D:\_dev\thethirdrace.com\node_modules\node-fetch\lib'

Import trace for requested module:
./node_modules/cross-fetch/dist/node-ponyfill.js
./node_modules/@supabase/postgrest-js/dist/module/lib/types.js
./node_modules/@supabase/postgrest-js/dist/module/index.js
./src/helper/Database.ts
./src/pages/api/payment/_middleware.api.ts

./node_modules/node-fetch/lib/index.mjs
Module not found: Can't resolve 'http' in 'D:\_dev\thethirdrace.com\node_modules\node-fetch\lib'

Import trace for requested module:
./node_modules/cross-fetch/dist/node-ponyfill.js
./node_modules/@supabase/postgrest-js/dist/module/lib/types.js
./node_modules/@supabase/postgrest-js/dist/module/index.js
./src/helper/Database.ts
./src/pages/api/payment/_middleware.api.ts

./node_modules/node-fetch/lib/index.mjs
Module not found: Can't resolve 'https' in 'D:\_dev\thethirdrace.com\node_modules\node-fetch\lib'

Import trace for requested module:
./node_modules/cross-fetch/dist/node-ponyfill.js
./node_modules/@supabase/postgrest-js/dist/module/lib/types.js
./node_modules/@supabase/postgrest-js/dist/module/index.js
./src/helper/Database.ts
./src/pages/api/payment/_middleware.api.ts


> You're using a Node.js module (querystring) which is not supported in the Edge Runtime.
Learn more: https://nextjs.org/docs/api-reference/edge-runtime
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Supabase packages use `cross-fetch`, which is not supported in the `middleware` runtime. To circumvent this, you can pass a custom `fetch` implementation when creating the `client`. For example:

```typescript
import { PostgrestClient, type PostgrestError } from '@supabase/postgrest-js'
import { supabaseAnonKey, supabaseUrl } from '~/helper/Authentication'

export const database = new PostgrestClient(`${supabaseUrl}/rest/v1`, {
	// Replace `cross-fetch` for `fetch` (the former doesn't work in Cloudflare Workers)
	fetch: (...args) => fetch(...args)
}).auth(supabaseAnonKey)
database.headers.apikey = supabaseAnonKey
```

This worked 100% in `12.1.5`, but `12.1.6` won't let me build at all.

After some research in the release notes, I think the problem is caused by the edge compiler from this PR: https://github.com/vercel/next.js/pull/36486

It would make sense the edge compiler doesn't expect node modules and only wants `fetch` given the environment is a Cloudflare Worker and not NodeJs. But at the same time, once actually compiled, the code will NOT use `cross-fetch`, it will use the normal `fetch`.  The problem is it's not letting me compile so I can give it what it wants... Chicken and egg problem...

Things that I've tried:
- I have tested with `canary`, got same problem
- I've done a stackblitz (linked below), it works, but uses babel instead of SWC. Unfortunately, I don't know how to make it use SWC and I have no idea why it works with Babel...
- I've commented the code using Supabase and the build starts working so I'm pretty confident this is the problem (also the error message kinda give it away)

### Expected Behavior

`yarn build` should be the project without error

### Link to reproduction

https://stackblitz.com/edit/nextjs-4vfysz?file=pages/api/_middleware.js

### To Reproduce

I can't reproduce it with Stackblitz because your template uses Babel instead of SWC.

I still gave a minimum reproduction through Stackblitz, but you'll need to run `yarn build` within a real repo that uses SWC as compiler.

Basically:

1. copy/paste `_middleware.ts` in a real repo under `page/api`
2. add dependencies to `@supabase/gotrue-js` and `@supabase/postgrest-js`
3. `yarn build`

**Comment by sannajammeh** [comment]
I ran into this. Ended up writing my own implementation of GoTrue using browser only deps to Get authentication working in middleware. Supabase is not treeshaking the server deps properly which I assume is what confuses this compiler.

**Comment by balazsorban44** [comment]
The local dev environment is intended to resemble the remote one to catch errors like this early. I think #37024 is more relevant here.

The problem is that `cross-fetch isn't loaded conditionally by `@supabase/postgrest-js` so it will always be imported, and since it uses Node.js specific APIs, it will cause the problem above.

The problem is this import:

https://github.com/supabase/postgrest-js/blob/b303b52c890f90c80e501a7b9160f588a38dc727/src/lib/types.ts#L1

Being used here:
https://github.com/supabase/postgrest-js/blob/b303b52c890f90c80e501a7b9160f588a38dc727/src/lib/types.ts#L70

Note that in your case you are using an Edge API route, so you wouldn't need to use Middleware here, see #37481.

**Comment by TheThirdRace** [comment]
@balazsorban44 

Thanks for your input, I'll check to open issues with Supabase.

Although, there's something that doesn't make sense for me with your last sentence.

To my knowledge, there's no such thing as an `Edge API route` in NextJs. You either have the good old `API route` running as a serverless function on AWS in exactly `1` data center OR you have `middleware` running on the `Edge` as a Cloudflare workers.

Is it possible this `Edge API route` is NOT mentioned at all in the documentation?  I can't find any mention of it anywhere beside NextJs team creating their own issues and PRs.

Could you elaborate a bit more on the subject please?

**Comment by balazsorban44** [comment]
See https://github.com/vercel/next.js/pull/37481. Documentation is coming soon. So as mentioned above, this is likely an issue that needs to be addressed in the packages you are using. I notified the Supabase team FYI.

**Comment by kiwicopple** [comment]
FYI, you can provide a custom fetch implementation for all supabase packages like [this](https://supabase.com/docs/reference/javascript/initializing#custom-fetch-implementation)

```js
import { createClient } from '@supabase/supabase-js'

const supabase = createClient('https://xyzcompany.supabase.co', 'public-anon-key', { 
  fetch: fetch.bind(globalThis) 
})
```

**Comment by TheThirdRace** [comment]
@kiwicopple 
I already do this, it simply doesn't work because Supabase imports directly `cross-fetch` before checking for a `fetch` replacement.

If Supabase were to use dynamic `import`, it would work most likely work.
